src/linux/sensor_ds18b20.c

```diff
@@ -217,10 +224,15 @@ ds18_send_and_request(struct ds18_s *d, uint32_t next_begin_time, uint8_t oid)
     if (d->status == W1_ERROR) {
         // try_shutdown expects a static string. Output the specific error,
         // then shut down with a generic error.
-        output("Error: %s", d->error);
         pthread_mutex_unlock(&d->lock);
-        try_shutdown("Error reading DS18B20 sensor");
-        return;
+        if (d->error_count > d->max_error_count) {
+            try_shutdown("Error reading DS18B20 sensor");
+            return;
+        } else {
+          sendf("ds18b20_result oid=%c next_clock=%u value=%i"
+                  , oid, next_begin_time, 0);
```

Comment:
What if 0 is configured to be out of range? We could check the range here, or base the value on the configured temperature range, but I think it would be simpler to wait and not send anything until after a successful read.

Reply:
Not sending a report is not ideal IMO because it completely hides the error. The current 0C solution at least shows it in mainsail. Of course, out of range is a problem.

What if we moved the error handling into python? The firmware could simply notify the host about the error without actually handling it. The notification would be easiest with a special temperature (like `0xFFFF`) or with a message type (`ds18b20_error`?).

The python code could then handle the error including out of range. This could be repeated for other sensors and the error handling could be more centralized in the future. Aggregate reports or nice visualizations would be possible.

Reply:
It does seem like reporting the error somehow would be useful even if the sensor recovers. Sending 0C so that the anomaly shows up in the temperature graph seems like an easy solution but maybe not the best case in the long term.

> This could be repeated for other sensors and the error handling could be more centralized in the future. Aggregate reports or nice visualizations would be possible.

I like that idea, but it sounds more ambitious than this fix. Are there other modules that report non-fatal errors now (and what do they do)?

Reply:
No other sensors yet, but plenty of discussion above. It seems that i2c sensors are similarly not 100% reliable on every read.

I agree 0°C is just a stopgap.

Reply:
> I like that idea, but it sounds more ambitious than this fix. Are there other modules that report non-fatal errors now (and what do they do)?

@joshhead See https://github.com/Klipper3d/klipper/issues/4874. A common fix would really make sense IMO

Reply:
FWIW, might be better to add a new field to the ds18b20_result message - for example, "ds18b20_result oid=%c next_clock=%u value=%i fault=%u".  See `thermocouple_response()` in src/thermocouple.c as an example.

-Kevin

Reply:
I have added the `fault` parameter.